File: src/figureIndentation.js
import _ from 'lodash'

const assert = (pred, msg) => {
  if (!pred) {
    throw new Error(msg)
  }
}

const basicChild = {
  getChildren: () => [],
  getBlock: () => [],
}

// Create nested object
const figureIndentation = (lines, indentation) => {
  const [line, nextLine, ...tail] = lines

  if (line === undefined) {
    return []
  }

  if (nextLine === undefined) {
    return [{
      line: line.line,
      getChildren: () => [],
      getBlock: () => [],
    }]
  }

  assert(line.indentation === indentation, 'death')
  assert(nextLine.indentation >= indentation, 'more death')

  if (nextLine.indentation === indentation) {
    return [{
      getChildren: () => [],
      getBlock: () => [],
      line: line.line,
    }].concat(figureIndentation([nextLine, ...tail], indentation))
  }

  const higherOrEqualIndentation =
    line => line.indentation >= nextLine.indentation

  const children = _.takeWhile([nextLine, ...tail], higherOrEqualIndentation)
  const newLines = _.dropWhile(tail, higherOrEqualIndentation)

  return [{
    line: line.line,
    getChildren: () => figureIndentation(children, nextLine.indentation),
    getBlock: () =>
      children
      .map(child => (' ').repeat(child.indentation - nextLine.indentation) + child.line)
  }].concat(figureIndentation(newLines, indentation))
}

export default figureIndentation

File: ruby/figureIndentation.rb
require_relative('classes')

def assert(pred, msg)
  if not pred
    raise Error, msg
  end
end

# Create nested object
def figureIndentation(lines, indentation)
  line, *first_tail = lines
  nextLine, *tail = first_tail

  if line.nil?
    return []
  end

  if nextLine.nil?
    return [Line.new(line.line)]
  end

  assert(line.indentation == indentation, 'death')
  assert(nextLine.indentation >= indentation, 'more death')

  if nextLine.indentation == indentation
    return [Line.new(line.line)] + figureIndentation(first_tail, indentation)
  end

  higherOrEqualIndentation = lambda {
    |line| line.indentation >= nextLine.indentation
  }

  children = first_tail.take_while(&higherOrEqualIndentation)
  newLines = first_tail.drop_while(&higherOrEqualIndentation)

  [Line.new(line.line, children)]
  .concat(figureIndentation(newLines, indentation))
end
